### PR TITLE
Trick: add a page to show a trick

### LINKED ISSUE #9: Add a page to show a trick

### PR DESCRIPTION
**Pull request by nicordev**
Will close #9 #10 #2 #3

### CODE REVIEW

**Comment by olix21** [comment]
Capture d’écran 2019-04-02 à 15 24 34

I think you could improve the css in this PR (trick title, edit/delete btns/images not centered, description)

File: src/Controller/HomeController.php [hunk 1]
```diff
@@ -2,16 +2,29 @@
...
 
+use App\Entity\Trick;
+use App\Helper\ControllerHelper;
+use Doctrine\Common\Persistence\ObjectManager;
 use Symfony\Bundle\FrameworkBundle\Controller\AbstractController;
 use Symfony\Component\Routing\Annotation\Route;
 
 class HomeController extends AbstractController
 {
+    public $tricksPerPage = 10;
```
Comment (olix21) [inline src/Controller/HomeController.php]: should be a private constant

File: src/Helper/ControllerHelper.php [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+class ControllerHelper
+{
+    /**
+     * Calculate the offset for paging
+     *
+     * @param int $page
+     * @param int $linesPerPage
+     * @return float|int
+     */
+    public static function getPagingOffset(int $page, int $linesPerPage)
```
Comment (olix21) [inline src/Helper/ControllerHelper.php]: Do you really need a Helper for that?

File: src/Controller/TrickController.php [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+        ]);
+    }
+
+    /**
+     * @Route("/create-trick", name="create_trick")
+     * @Route("/edit-trick/{id}", name="edit_trick", requirements={"id": "\d+"})
+     */
+    public function createOrEdit(Request $request, ObjectManager $manager, Trick $trick = null)
+    {
+        if (!$trick) {
```
Comment (olix21) [inline src/Controller/TrickController.php]: `
$trick = $trick ?? new Trick();
`
could also work

File: src/DataFixtures/AppFixtures.php [hunk 4]
```diff
@@ -0,0 +1,17 @@
+<?php
+
+namespace App\DataFixtures;
+
+use Doctrine\Bundle\FixturesBundle\Fixture;
+use Doctrine\Common\Persistence\ObjectManager;
+
+class AppFixtures extends Fixture
```
Comment (olix21) [inline src/DataFixtures/AppFixtures.php]: remove as it's unused

File: src/Controller/HomeController.php (hunk 1 above)
Comment (nicordev) [reply src/Controller/HomeController.php]: Done.

File: src/Helper/ControllerHelper.php (hunk 2 above)
Comment (nicordev) [reply src/Helper/ControllerHelper.php]: It's just the beginning of a Symfony service to help me with paging. It's in progress, I'm gonna improve this.

File: src/Controller/TrickController.php (hunk 3 above)
Comment (nicordev) [reply src/Controller/TrickController.php]: Done.

File: src/DataFixtures/AppFixtures.php (hunk 4 above)
Comment (nicordev) [reply src/DataFixtures/AppFixtures.php]: Done.

File: src/Helper/ControllerHelper.php (hunk 2 above)
Comment (nicordev) [reply src/Helper/ControllerHelper.php]: And it's done! I now have a Paginator which saves me from writing duplicated code as in the previous project.

File: demodata/getYouTubeUrls.js [hunk 5]
```diff
@@ -0,0 +1,42 @@
+/*
+ * Code à executer dans une ardoise javascript ici : https://www.youtube.com/results?search_query=snowboard+tricks
```
Comment (olix21) [inline demodata/getYouTubeUrls.js]: I don't understand the goal of this file nor the `demodata/youtube_videos` file.
If you want to display some content, why don't you create some `Trick` fixtures?

File: src/Controller/HomeController.php [hunk 6]
```diff
@@ -2,16 +2,38 @@
...
 
 class HomeController extends AbstractController
 {
+    public const TRICKS_PER_PAGE = 10;
+
     /**
+     * Show the home page
+     *
      * @Route("/", name="home")
+     * @Route("/home/{page}", name="home_paging", requirements={"page": "\d+"})
```
Comment (olix21) [inline src/Controller/HomeController.php]: Why not, but you can also handle the pagination without creating a dedicated route:

`
$router->generate('home', ['page' => 10])
// As the route has no `page` parameter, the router will pass this parameter in the query parameters (mywebsite.com/my_route?page=10)

// Controller
$page = $request->query->getInt('page', 1)
`

File: src/Controller/TrickController.php [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+    /**
+     * @Route("/delete-trick/{id}", name="delete_trick", requirements={"id": "\d+"})
+     */
+    public function delete(ObjectManager $manager, Trick $trick)
+    {
+        $trickName = $trick->getName();
+        $manager->remove($trick);
+        $manager->flush();
+
+        return $this->redirectToRoute("home", ["message" => "Le trick $trickName a été supprimé"]);
```
Comment (olix21) [inline src/Controller/TrickController.php]: Use the FlashMessage feature from Symfony

File: src/Controller/TrickController.php [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+    {
+        $trick = $trick ?? new Trick();
+
+        $form = $this->createForm(TrickType::class, $trick);
+
+        $form->handleRequest($request);
+
+        if ($form->isSubmitted() && $form->isValid()) {
+            if (!$trick->getId()) {
+                $trick->setCreatedAt(new DateTime);
```
Comment (olix21) [inline src/Controller/TrickController.php]: Move this assignation is the Trick constructor

File: src/Controller/TrickController.php [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+            } else {
+                $trick->setModifiedAt(new DateTime);
```
Comment (olix21) [inline src/Controller/TrickController.php]: You can use the Doctrine Entity LifeCycle to update the Modification Date

File: demodata/getYouTubeUrls.js (hunk 5 above)
Comment (nicordev) [reply demodata/getYouTubeUrls.js]: I did create a TrickFixtures which get youtube videos urls from this file.
What's wrong? Should I move this file into the DataFixtures folder?

File: src/Controller/TrickController.php (hunk 8 above)
Comment (nicordev) [reply src/Controller/TrickController.php]: Better: I use Doctrine PrePersist annotation now!

File: src/Controller/TrickController.php (hunk 9 above)
Comment (nicordev) [reply src/Controller/TrickController.php]: Yes, true! I didn't know this feature, awesome!

File: src/Controller/TrickController.php (hunk 7 above)
Comment (nicordev) [reply src/Controller/TrickController.php]: Great advice! It works well. Thanks!

File: src/Controller/HomeController.php (hunk 6 above)
Comment (nicordev) [reply src/Controller/HomeController.php]: Ok, so I merged the 2 routes into one : `"/{page}"`

**Pull request merged**